libraries/package-extractor/src/AssetHandler.ts

```diff
@@ -0,0 +1,224 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import path from 'node:path';
+import fs from 'node:fs';
+import { Async, FileSystem, Path, type FileSystemStats } from '@rushstack/node-core-library';
+import type { ITerminal } from '@rushstack/terminal';
+import { ArchiveManager } from './ArchiveManager';
+import type { IExtractorOptions, LinkCreationMode } from './PackageExtractor';
+import type { ILinkInfo, SymlinkAnalyzer } from './SymlinkAnalyzer';
+import { remapSourcePathForTargetFolder } from './Utils';
+
+export interface IIncludeAssetOptions {
+  sourceFilePath?: string;
+  sourceFileStats?: FileSystemStats;
+  sourceFileContent?: string | Buffer;
+  targetFilePath: string;
+  ignoreIfExisting?: boolean;
+}
+
+export interface IIncludeAssetPathOptions extends IIncludeAssetOptions {
+  sourceFilePath: string;
+  sourceFileContent?: never;
+}
+
+export interface IIncludeExistingAssetPathOptions extends IIncludeAssetOptions {
+  sourceFilePath?: never;
+  sourceFileContent?: never;
+}
+
+export interface IIncludeAssetContentOptions extends IIncludeAssetOptions {
+  sourceFileContent: string | Buffer;
+  sourceFilePath?: never;
+  sourceFileStats?: never;
+}
+
+export interface IAssetHandlerOptions extends IExtractorOptions {
+  symlinkAnalyzer: SymlinkAnalyzer;
+}
+
+export class AssetHandler {
+  private readonly _terminal: ITerminal;
+  private readonly _sourceRootFolder: string;
+  private readonly _targetRootFolder: string;
+  private readonly _createArchiveOnly: boolean;
+  private readonly _symlinkAnalyzer: SymlinkAnalyzer;
+  private readonly _archiveManager: ArchiveManager | undefined;
+  private readonly _archiveFilePath: string | undefined;
+  private readonly _linkCreationMode: LinkCreationMode;
+  private readonly _includedAssetPaths: Set<string> = new Set<string>();
+  private _isFinalized: boolean = false;
+
+  public constructor(options: IAssetHandlerOptions) {
+    const {
+      terminal,
+      sourceRootFolder,
+      targetRootFolder,
+      linkCreation,
+      symlinkAnalyzer,
+      createArchiveFilePath,
+      createArchiveOnly = false
+    } = options;
+    this._terminal = terminal;
+    this._sourceRootFolder = sourceRootFolder;
+    this._targetRootFolder = targetRootFolder;
+    this._symlinkAnalyzer = symlinkAnalyzer;
+    if (createArchiveFilePath) {
+      if (path.extname(createArchiveFilePath) !== '.zip') {
+        throw new Error('Only archives with the .zip file extension are currently supported.');
+      }
```

Comment:
It'd be great to add support for tarballs too.

Reply:
Agree, though I think that would be a future PR/out of scope for this one.